test/xgo_integration/shadow_test.go

```diff
@@ -0,0 +1,93 @@
+package xgo_integration
+
+import (
+	"os"
+	"os/exec"
+	"path/filepath"
+	"testing"
+
+	"github.com/xhd2015/xgo/support/cmd"
+	"github.com/xhd2015/xgo/support/osinfo"
+)
+
+func checkedShadow(t *testing.T) string {
+	checkXgo(t)
+
+	shadow, err := cmd.Output("xgo", "shadow")
+	if err != nil {
+		t.Fatal(err)
+	}
+	if shadow == "" {
+		t.Fatalf("expect shadow path printed, actual nothing")
+	}
+	return shadow
+}
+func TestShadowPrintPath(t *testing.T) {
+	shadow := checkedShadow(t)
+	shadowGo := filepath.Join(shadow, "go"+osinfo.EXE_SUFFIX)
+	stat, err := os.Stat(shadowGo)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if stat.IsDir() {
+		t.Fatalf("expect shadow go to be file, actual: dir, %s", shadowGo)
+	}
+}
+
+func TestShadowPathLookupGo(t *testing.T) {
+	shadow := checkedShadow(t)
+	goBinary, err := exec.LookPath("go")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	shadowGo := filepath.Join(shadow, "go"+osinfo.EXE_SUFFIX)
+	if goBinary == shadowGo {
+		t.Skipf("skipped because shadow on PATH")
+	}
+	goroot, err := cmd.Output(goBinary, "env", "GOROOT")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	gorootGo := filepath.Join(goroot, "bin", "go"+osinfo.EXE_SUFFIX)
+	if gorootGo != goBinary {
+		t.Fatalf("go not inside GOROOT: %s", goBinary)
+	}
+	var goAfterShadow string
+	oldPath := os.Getenv("PATH")
+	withPathEnv(shadow+string(filepath.ListSeparator)+oldPath, func() {
```

Comment:
This should be a filepath.Join I think

Reply:
filepath.Join uses filepath.Separator, while here is filepath.ListSeapartor

Reply:
Oh, that's right